src/tools/illink/src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -1897,11 +1886,47 @@ internal void MarkMethodVisibleToReflection (MethodReference method, in Dependen
 			}
 		}
 
+		bool MarkMethodsVisibleToReflection (TypeDefinition type, in DependencyInfo reason, MessageOrigin origin)
+		{
+			if (!type.HasMethods)
+				return false;
+
+			foreach (MethodDefinition method in type.Methods) {
+				if (IsFullyPreservedAction (Annotations.GetAction (type.Module.Assembly)))
+					Annotations.SetAction (method, MethodAction.ForceParse);
+				MarkMethodVisibleToReflection (method, reason, origin);
+			}
+			return true;
+		}
+
 		internal void MarkFieldVisibleToReflection (FieldReference field, in DependencyInfo reason, in MessageOrigin origin)
 		{
 			MarkField (field, reason, origin);
 		}
 
+		bool MarkFieldsVisibleToReflection (TypeDefinition type, in DependencyInfo reason, MessageOrigin origin, bool markBackingFieldsOnlyIfPropertyMarked = false)
+		{
+			if (!type.HasFields)
+				return false;
+
+			foreach (FieldDefinition field in type.Fields) {
+				if (markBackingFieldsOnlyIfPropertyMarked && field.Name.EndsWith (">k__BackingField", StringComparison.Ordinal)) {
+					// We can't reliably construct the expected property name from the backing field name for all compilers
+					// because csc shortens the name of the backing field in some cases
+					// For example:
+					// Field Name = <IFoo<int>.Bar>k__BackingField
+					// Property Name = IFoo<System.Int32>.Bar
+					//
+					// instead we will search the properties and find the one that makes use of the current backing field
+					var propertyDefinition = SearchPropertiesForMatchingFieldDefinition (field);
+					if (propertyDefinition != null && !Annotations.IsMarked (propertyDefinition))
+						continue;
```

Comment:
Do we ever need to mark a backing field as visible to reflection since it's reflection over compiler generated code? Could we remove the IsMarked check?